File: 剑指Offer/22.链表中倒数第k个节点.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* getKthFromEnd(ListNode* head, int k) {
        if (head == NULL) return NULL;
        if (k == 0) return NULL;
        ListNode *fast = head, *slow = head;
        while (k--){
            if (fast == NULL) return NULL;
            fast = fast->next;
        }
        while (fast != NULL) {
            fast = fast->next;
            slow = slow->next;
        }
        return slow;
    }
};

File: 剑指Offer/22.链表中倒数第k个节点.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getKthFromEnd(self, head: ListNode, k: int) -> ListNode:
        if head is None:
            return None
        if k == 0:
            return None

        fast = head;
        slow = head;
        while k != 0:
            if fast is None:
                return None
            fast = fast.next
            k -= 1

        while fast is not None:
            fast = fast.next
            slow = slow.next

        return slow